### PR TITLE
task/18342

### PR DESCRIPTION
**Pull request by irl**

### CODE REVIEW

File: src/main/java/org/torproject/onionoo/util/DomainNameSystem.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package org.torproject.onionoo.util;
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/util/DomainNameSystem.java:4]: License blurb is missing.

File: src/main/java/org/torproject/onionoo/util/DomainNameSystem.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+public class DomainNameSystem {
+
+  private static DirContext dnsContext;
+  private static Logger log = LoggerFactory.getLogger(DomainNameSystem.class);
+
+  static {
+    Hashtable<String, String> envProps = new Hashtable<String, String>();
+    envProps.put(Context.INITIAL_CONTEXT_FACTORY,
+            "com.sun.jndi.dns.DnsContextFactory");
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/util/DomainNameSystem.java]: Use 4 spaces for indenting here. Run `ant checks` to check things like this. (Not 100% sure whether this shows up there, I didn't try.) There are more places where more whitespace is used than necessary, and I'm not marking any subsequent places.

**Review by kloesing — CHANGES_REQUESTED** [review]
Looks good overall! I'll have more feedback on the third commit in a bit.

File: src/main/java/org/torproject/onionoo/docs/DetailsDocument.java [hunk 3]
```diff
@@ -242,6 +242,74 @@ public String getHostName() {
...
 
+  private List<String> verified_host_names;
+
+  /**
+   * Creates a copy of the list with each string escaped for JSON compatibility
+   * and sets this as the verified host names, unless the argument was null in
+   * which case the verified host names are just set to null.
+   */
+  public void setVerifiedHostNames(List<String> verifiedHostNames) {
+    if (verifiedHostNames == null) {
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/docs/DetailsDocument.java]: We're trying to write checks like this the other way around, that is, `null == verifiedHostNames`, to minimize the risk of writing an assignment (`=`) where we really want a comparison (`==`). So, please just swap things here and in subsequent places. (There's more reasoning on the internets if you look for it.)

File: src/main/java/org/torproject/onionoo/docs/DetailsStatus.java [hunk 4]
```diff
@@ -531,6 +532,75 @@ public String getHostName() {
...
+   * Creates a copy of the list with each string escaped for JSON compatibility
+   * and sets this as the unverified host names, unless the argument was null in
+   * which case the unverified host names are just set to null.
+   */
+  public void setUnverifiedHostNames(List<String> unverifiedHostNames) {
+    if (unverifiedHostNames == null) {
+      this.unverified_host_names = null;
+      return;
+    }
+    System.out.println(unverifiedHostNames.toString());
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/docs/DetailsStatus.java]: This line doesn't belong here. Maybe search for similar lines, too, I just spotted this one and didn't search using `grep`.

File: src/main/java/org/torproject/onionoo/docs/DocumentStore.java [hunk 5]
```diff
@@ -11,7 +11,6 @@
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
-
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/docs/DocumentStore.java]: Please avoid unnecessary whitespace changes, if possible. I know, it was more likely the IDE, not you, but maybe you can teach it to not mess with imports unless instructed so.

File: src/main/java/org/torproject/onionoo/docs/SummaryDocument.java [hunk 6]
```diff
@@ -309,6 +309,28 @@ public String getHostName() {
     return this.hostName;
   }
 
+  @JsonProperty("vh")
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/docs/SummaryDocument.java:312]: Hmm, hmm, hmm, do we really want to include these new fields in summary documents, too? Maybe, I didn't think too hard about this yet. What do you think, should this change be here or not? (Truly open question.)

File: src/main/java/org/torproject/onionoo/docs/SummaryDocument.java [hunk 7]
```diff
@@ -309,6 +309,28 @@ public String getHostName() {
...
+  }
+
+  @JsonProperty("uh")
+  private List<String> unverifiedHostNames;
+
+  public void setUnverifiedHostNames(List<String> unverifiedHostNames) {
+    this.unverifiedHostNames = unverifiedHostNames;
+  }
+
+  public List<String> getUnerifiedHostNames() {
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/docs/SummaryDocument.java]: Lacks a v. Maybe search for similar typos in this patch.

File: src/main/java/org/torproject/onionoo/server/NodeIndexer.java [hunk 8]
```diff
@@ -266,8 +268,18 @@ private void indexNodeStatuses() {
         newRelaysByVersion.get(version).add(fingerprint);
         newRelaysByVersion.get(version).add(hashedFingerprint);
       }
-      String hostName = entry.getHostName();
-      if (null != hostName) {
+      List<String> allHostNames;
+      try {
+        allHostNames = new ArrayList<>(entry.getVerifiedHostNames());
+      } catch (NullPointerException e) {
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/server/NodeIndexer.java]: Catching `NullPointerException`s is so Python. Better check for `null` first and you can easily avoid catching this exception at all. Also applies for subsequent code.

File: src/main/java/org/torproject/onionoo/updater/NodeDetailsStatusUpdater.java [hunk 9]
```diff
@@ -757,18 +757,39 @@ private void computeEffectiveAndExtendedFamilies() {
...
+              verifiedLookupResults.get(nodeStatus.getAddress());
+      List<String> unverifiedHostNames =
+              unverifiedLookupResults.get(nodeStatus.getAddress());
       if (hostName != null) {
         this.rdnsLookupResults.put(fingerprint, hostName);
+      } else {
+        /* Maintains bug compatibility with previous implementation */
+        this.rdnsLookupResults.put(fingerprint, nodeStatus.getAddress());
+      }
+      if (verifiedHostNames != null && verifiedHostNames.size() > 0) {
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/updater/NodeDetailsStatusUpdater.java]: See above, please rewrite as `null != verifiedHostNames`, mostly for consistency.

Comment (kloesing) [reply src/main/java/org/torproject/onionoo/updater/NodeDetailsStatusUpdater.java]: Ah, and rather than checking `xy.size() > 0`, simply check `!xy.isEmpty()`. Easier to read.

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java [hunk 10]
```diff
@@ -23,7 +27,6 @@
   public RdnsLookupRequest(
       ReverseDomainNameResolver reverseDomainNameResolver,
       RdnsLookupWorker parent, String address) {
-    this.reverseDomainNameResolver = reverseDomainNameResolver;
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: I wonder, should we simply pass a `DomainNameSystem` reference here instead? Right now it's implemented as a singleton, right? That is hard to test then.

Somewhat unrelated, the class name `DomainNameSystem` is a bit ambiguous. Can't we keep the `ReverseDomainNameResolver` and just rewrite the contained code? At least that class name says what the class does/is responsible for.

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java [hunk 11]
```diff
@@ -32,23 +35,64 @@ public RdnsLookupRequest(
   public void run() {
     this.lookupStartedMillis = System.currentTimeMillis();
     try {
-      String result = InetAddress.getByName(this.address).getHostName();
+      final List<String> verified_results = new ArrayList<>();
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: Did this code pass `ant checks`? For local variables, avoid snake case. The only place where we use that is for old classes that are serialized using Jackson that we didn't update yet. Otherwise, avoid snake case.

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java [hunk 12]
```diff
@@ -32,23 +35,64 @@ public RdnsLookupRequest(
...
+      /* The Onionoo field is omitted for both lookup failure and absence of
+       * a host name. We'll try again the next time. */
     }
     this.lookupCompletedMillis = System.currentTimeMillis();
     this.parent.interrupt();
   }
 
   public synchronized String getHostName() {
-    return hostName;
+    if (verifiedHostNames == null ) {
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: Please always refer to members using `this.`. It is redundant, but it's more explicit when reading the code, and it's consistent with the current code.

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java [hunk 13]
```diff
@@ -32,23 +35,64 @@ public RdnsLookupRequest(
...
+      return null;
+    } else if (verifiedHostNames.size() > 0) {
+      return verifiedHostNames.get(0);
+    } else {
+      return null;
+    }
```
Comment (kloesing) [inline src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: Two cases have the exact same code block here. You could easily combine them by saying `if (null != verifiedHostNames && !verifiedHostNames.isEmpty())`.

**Review by kloesing — CHANGES_REQUESTED** [review]
Couple of findings in this commit. Some things might seem nitpicky, but it's better to discuss them now and have a better common understanding of our code should look like.

Apart from all those comments, great work! Really looking forward to getting this into the next release. Thanks!

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java (hunk 11 above)
Comment (irl) [reply src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: It did pass `ant checks`. I had made sure that everything passed before pushing.

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java (hunk 10 above)
Comment (irl) [reply src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: `ReverseDomainNameResolver` is not really the resolver, but just a part of the updater process that manages requests and workers. `DomainNameSystem` is currently implemented with a static method.

File: src/main/java/org/torproject/onionoo/docs/SummaryDocument.java (hunk 6 above)
Comment (irl) [reply src/main/java/org/torproject/onionoo/docs/SummaryDocument.java:312]: They need to be here to allow for filtering based on them.

File: src/main/java/org/torproject/onionoo/util/DomainNameSystem.java (hunk 1 above)
Comment (irl) [reply src/main/java/org/torproject/onionoo/util/DomainNameSystem.java:4]: Added this.

Comment (kloesing) [reply src/main/java/org/torproject/onionoo/util/DomainNameSystem.java:4]: I usually only include the current year in newly created files, but I'm also fine with including several years. Why 2013 though? (Mainly for future reference.)

File: src/main/java/org/torproject/onionoo/docs/SummaryDocument.java (hunk 6 above)
Comment (kloesing) [reply src/main/java/org/torproject/onionoo/docs/SummaryDocument.java:312]: Alright, makes sense. (I didn't look at this code in a while.)

File: src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java (hunk 10 above)
Comment (kloesing) [reply src/main/java/org/torproject/onionoo/updater/RdnsLookupRequest.java]: This is the only remaining concern I have: does `DomainNameSystem` have to be implemented with a static method? This effectively makes the class a singleton. Do you mind changing the static block to a constructor and the static method to an instance method? Otherwise this may bite us in the future.

File: src/main/java/org/torproject/onionoo/util/DomainNameSystem.java (hunk 1 above)
Comment (irl) [reply src/main/java/org/torproject/onionoo/util/DomainNameSystem.java:4]: Ah ok. Usually I've used copyright years on a per-project basis, not per-file. Will change to just 2018 for consistency.
